### DISCUSSION TITLE
How should I handle disposed references?

### DISCUSSION THREAD

**Discussion by Maldus512** [rrousselGit/riverpod#3438] [Q&A] [answered]
This question is similar to one I've already asked (https://github.com/rrousselGit/riverpod/discussions/3043), but more general now.

I frequently find myself splitting provider access along my Widget tree; I will have multiple ConsumerWidgets that each watch a specific provider or parts of it to ensure minimal refresh on data change. If I understand correctly, this is how I'm supposed to work instead of passing data and callbacks down the tree.
However, issues arise when a Widget deep down the tree manipulates my state in such a way that it is not displayed anymore and disposed of, thus ending up with an invalid WidgetRef in the middle of an async callback.

As per my previous question, one can simply save a reference to the required provider and use that instead of requesting it again with an invalid ref. However this doesn't solve every problem: I've recently needed to invalidate a FutureProvider after the widget was disposed of, but since it requires calling `ref.invalidate` I'm stuck. 

Here is some pseudocode for reference:

```
            onClick: () async {
              final pageStateNotifier = ref.read(_pageStateProvider.notifier);

              pageStateNotifier.loading(); // <- Here the widget disappears to make way to a loading screen
              await networkOperation();
              ref.invalidate(_systemsProvider(this.deviceId)); // <- invalid reference
              pageStateNotifier.loadingDone();
            }),

```

I have two questions:

 1. Is there a way to avoid this issue specifically? That is, refresh a FutureProvider when the current ref I'm holding is invalid?
 2. Am I doing things correctly? Is there a different best practice to follow to avoid this errors? When it comes to using providers on a non trivial Widget tree I feel like I'm randomly stitching together references in order to make sure everything works as expected.

**Comment by rrousselGit** [comment, marked as answer]
Sounds like all of your logic should be a method on `_pageStateProvider` instead.

Make it a Notifier and have your logic there.
